presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PredicatePushDown.java

```diff
@@ -447,15 +452,24 @@ private OuterJoinPushDownResult processLimitedOuterJoin(Expression inheritedPred
                 }
             }
 
-            // See if we can push down any outer or join predicates to the inner side
-            for (Expression conjunct : EqualityInference.nonInferrableConjuncts(and(outerEffectivePredicate, joinPredicate))) {
+            // See if we can push down any outer predicates to the inner side
+            for (Expression conjunct : EqualityInference.nonInferrableConjuncts(outerEffectivePredicate)) {
                 Expression rewritten = potentialNullSymbolInference.rewriteExpression(conjunct, not(in(outerSymbols)));
                 if (rewritten != null) {
                     innerPushdownConjuncts.add(rewritten);
                 }
             }
 
-            // TODO: consider adding join predicate optimizations to outer joins
+            // See if we can push down any join predicates to the inner side
+            for (Expression conjunct : extractConjuncts(joinPredicate)) {
+                Expression rewritten = innerInference.rewriteExpression(conjunct, not(in(outerSymbols)));
```

Comment:
We discussed this code a lot here today. 

Actually it seems you cannot use `potentialNullSymbolInference` here if you are removing the conjunct from `joinConditions` after rewrite. That way you loose information from join clause which is important for proper join semantics.

Extreme example:

```
   LOJ(a=b)
      TS(t1(a))
      TS(t2(b)
```

cannot be rewritten to:

```
   LOJ()
      TS(t1(a))
      FILTER(b=b)
          TS(t2(b))
```

And that is exact behaviour I was getting when I just replaced `innerInference` with `potentialNullSymbolInference`.
Potentially we could use `potentialNullSymbolInference` for extra rewrites which does not remove conjuncts from joinClause and do just pushDown. But I did not want to change that in this patch to be as safe as possible in bugfix PR.

The added commit prevents removal conjuncts from original join condition in case when the original conjunct touched outer symbols. This is just to prevent from getting empty condition in `LOJ` which would imply slow execution currently.
I have second thoughts on that. Probably the better approach would be to leave any condition which can be executed as fast-copiled LookupJoin in join clauses until we have proper implementation of NestedLoopsJoin.
And even then it would probably be not that obvious if removing conjuncts from `joinClause` is always good choice.
As it could for example make data distribution during execution worse.

I will be back by the computer in 2-3 hours. Let me know what you think.

cc: @sogorkis, @pnowojski